File: Codeforces/D_Grid_and_Magnet.cpp
#include <bits/stdc++.h>
using namespace std;

#define DEBUG(x...) { cout << "(" << #x << ")" << " = ( "; Print(x); }
template <typename T1> void Print(T1 t1) { cout << t1 << " )" << endl; }
template <typename T1, typename... T2>
void Print(T1 t1, T2... t2) { cout << t1 << " , "; Print(t2...); }
 

#define int long long
#define pi (3.141592653589)
#define mod 1000000007
#define float double
#define ff first
#define ss second
#define mk make_pair
#define pb push_back
#define rep(i, start, end) for (int i = start; i < end; i++)
#define ld long double
#define fast ios_base::sync_with_stdio(false), cin.tie(nullptr), cout.tie(nullptr);
int inf = 1000000000000000000;
using ii = pair<int, int>;
int n,m;


vector<int> dx = {1,0,-1,0};
vector<int> dy = {0,1,0,-1};

int bfs(int x, int y, vector<vector<char>>&vec,vector<vector<int>>&magnets,vector<vector<int>>&vis){
    // cout<<"bsf "<<x<<" "<<y<<endl;
    queue<vector<int>> q;
    set<int> st;
    vis[x][y]=1;
    q.push({x,y});
    st.insert(x*m+y);
    while(!q.empty()){
        auto v = q.front();
        q.pop();
        auto xx =v[0],yy=v[1];

        for (int i=0;i<4;i++){
            int nx = xx+dx[i];
            int ny = yy+dy[i];
            if (nx<0 || nx>=n || ny<0 || ny>=m) continue;
            // cout<<nx<<" bfs call "<<ny<<endl;
            int nxy = nx*m+ny;
            if (st.count(nxy)>0) continue;
            st.insert(nxy);
            vis[nx][ny]=1;
            if (magnets[nx][ny]==1) continue;
            q.push({nx,ny});


        }

    }
    return st.size();

}
    
void solve()
{
    cin>>n>>m;
    vector<vector<char>> vec(n, vector<char>(m));
    vector<vector<int>> magnets(n,vector<int>(m,0));
    vector<vector<int>> vis(n,vector<int>(m,0));
    
    rep(i,0,n){
        rep(j,0,m){
            cin>>vec[i][j];
        }
    }

    rep(i,0,n){
        rep(j,0,m){
            if (vec[i][j]=='#'){
                magnets[i][j]=1;
                // cout<<i<<" "<<j</<endl;
                rep(k,0,4){
                    int ni=i+dx[k], nj=j+dy[k];
                    if (ni<0 || ni>=n || nj<0 || nj>=m) continue;
                    // cout<<ni<<" "<<nj<<" ...."<<endl;
                    magnets[ni][nj]=1;
                }
            }
        }
    }
    
    

    // rep(i,0,n){
    //     rep(j,0,m){
    //         cout<<magnets[i][j]<<' ';
    //     }
    //     cout<<endl;
    // }
    // cout<<endl;

    int ans=1;
    rep(i,0,n){
        rep(j,0,m){
            if (vis[i][j]==1) continue;
            if (magnets[i][j]==1) continue;
            // cout<<bfs(i,j,vec,magnets,vis)<<" "<<i<<" "<<j<<endl;
            ans = max(ans, bfs(i,j,vec,magnets,vis));
            }
        }

    cout<<ans<<endl;

}

signed main()
{
    fast
    // int t; cin>>t; while(t--)

    #ifndef ONLINE_JUDGE
    freopen("input.txt","r",stdin);
    freopen("output.txt","w",stdout);
    #endif
    solve();
    return 0;
}

File: Codeforces/D_Grid_and_Magnet.py
import sys

# sys.setrecursionlimit(1000005)
# sys.set_int_max_str_digits(200005)
int1 = lambda x: int(x)-1
pDB = lambda *x: print(*x, end="\n", file=sys.stderr)
p2D = lambda x: print(*x, sep="\n", end="\n\n", file=sys.stderr)
def II(): return int(sys.stdin.readline())
def LI(): return list(map(int, sys.stdin.readline().split()))
def LLI(rows_number): return [LI() for _ in range(rows_number)]
def LI1(): return list(map(int1, sys.stdin.readline().split()))
def LLI1(rows_number): return [LI1() for _ in range(rows_number)]
def SI(): return sys.stdin.readline().rstrip()
from os import path

if (path.exists('input.txt')):
    sys.stdin  = open('input.txt', 'r')  
    sys.stdout = open('output.txt', 'w') 


dij = [(0, 1), (-1, 0), (0, -1), (1, 0)]
# dij = [(0, 1), (-1, 0), (0, -1), (1, 0), (1, 1), (1, -1), (-1, 1), (-1, -1)]
# inf = -1-(-1 << 31)
inf = -1-(-1 << 62)

# md = 10**9+7
md = 998244353

def dfs(i,j):
    vis[i][j]=1
    st=[(i,j)]
    ij=set()
    ij.add(i*w+j)
    while st:
        i,j=st.pop()
        for di, dj in dij:
            ni, nj = i+di, j+dj
            if ni < 0 or nj < 0 or ni >= h or nj >= w: continue
            nij=ni*w+nj
            if nij in ij:continue
            ij.add(nij)
            vis[ni][nj]=1
            if mag[ni][nj]:continue
            st.append((ni,nj))
    return len(ij)

h,w=LI()
ss=[SI() for _ in range(h)]
mag=[[0]*w for _ in range(h)]
vis = [[0]*w for _ in range(h)]
for i in range(h):
    for j in range(w):
        if ss[i][j]=="#":
            mag[i][j]=1
            for di, dj in dij:
                ni, nj = i+di, j+dj
                if ni < 0 or nj < 0 or ni >= h or nj >= w: continue
                mag[ni][nj]=1
ans=1
for i in range(h):
    for j in range(w):
        if mag[i][j]:continue
        if vis[i][j]:continue
        print(i,j, dfs(i,j))
        print(vis)
        ans=max(ans,dfs(i,j))

print(ans)
